src/main/java/rx/internal/operators/OperatorPublish.java

```diff
@@ -15,367 +15,736 @@
+    /** The source observable. */
     final Observable<? extends T> source;
-    private final RequestHandler<T> requestHandler;
+    /** Holds the current subscriber that is, will be or just was subscribed to the source observable. */
+    final AtomicReference<PublishSubscriber<T>> current;
 
+    /**
+     * Creates a OperatorPublish instance to publish values of the given source observable.
+     * @param source the source observable
+     * @return the connectable observable
+     */
     public static <T> ConnectableObservable<T> create(Observable<? extends T> source) {
-        return new OperatorPublish<T>(source);
+        // the current connection to source needs to be shared between the operator and its onSubscribe call
+        final AtomicReference<PublishSubscriber<T>> curr = new AtomicReference<PublishSubscriber<T>>();
+        OnSubscribe<T> onSubscribe = new OnSubscribe<T>() {
+            @Override
+            public void call(Subscriber<? super T> child) {
+                // concurrent connection/disconnection may change the state, 
+                // we loop to be atomic while the child subscribes
+                for (;;) {
+                    // get the current subscriber-to-source
+                    PublishSubscriber<T> r = curr.get();
+                    // if there isn't one or it is unsubscribed
+                    if (r == null || r.isUnsubscribed()) {
+                        // create a new subscriber to source
+                        PublishSubscriber<T> u = new PublishSubscriber<T>(curr);
+                        // perform extra initialization to avoid 'this' to escape during construction
+                        u.init();
+                        // let's try setting it as the current subscriber-to-source
+                        if (!curr.compareAndSet(r, u)) {
+                            // didn't work, maybe someone else did it or the current subscriber 
+                            // to source has just finished
+                            continue;
+                        }
+                        // we won, let's use it going onwards
+                        r = u;
+                    }
+                    
+                    // create the backpressure-managing producer for this child
+                    InnerProducer<T> inner = new InnerProducer<T>(r, child);
+                    /*
+                     * Try adding it to the current subscriber-to-source, add is atomic in respect 
+                     * to other adds and the termination of the subscriber-to-source.
+                     */
+                    if (!r.add(inner)) {
+                        /*
+                         * The current PublishSubscriber has been terminated, try with a newer one.
+                         */
+                        continue;
+                        /*
+                         * Note: although technically corrent, concurrent disconnects can cause 
```

Comment:
By concurrent disconnect, I assume you mean concurrent subscribe/disconnect? If so then yes, the child subscribers may never receive anything if it is not re-connected. But that is exactly correct behavior if that's what happens. 

Am I understanding your comment here correctly?